server/src/main/java/org/opensearch/plugins/SystemIndexPlugin.java

```diff
@@ -55,4 +58,13 @@ public interface SystemIndexPlugin extends ActionPlugin {
     default Collection<SystemIndexDescriptor> getSystemIndexDescriptors(Settings settings) {
         return Collections.emptyList();
     }
+
+    /**
+     * This function passes the registered system indices to a plugin.
+     *
+     * Note: Only one installed plugin may implement onSystemIndices.
+     */
+    default Consumer<Map<String, Set<String>>> onSystemIndices() {
```

Comment:
@cwperks it looks like we are opening the gate to every single plugin to list system indices which I think should be protected by permissions, no?

Reply:
@reta no, this extension point is limited to a single plugin. It follows the same pattern as [ActionPlugin.getRestHandlerWrapper](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/ActionPlugin.java#L130-L153) which the security plugin uses to perform auth.

Reply:
@andrross brought up a good point here: https://github.com/opensearch-project/security/issues/4439#issuecomment-2174492217

I replied about the reasons for passing these to the security plugin, but curious on others thought as well?

Reply:
Another question I have is related to API:
 - why we use `Map<String, Set<String>>` and not `SystemIndexDescriptor` fe?
 - why we use `Consumer` instead of simpler hook like (like other APIs, onIndexModule, onNodeStarted, ...)?
   ```
  default void onSystemIndex(SystemIndexDescriptor descriptor) {
    ...
  }
   ```

Reply:
Agree on the API signature, but on the surface this PR seems like we're exposing something that is pretty much exclusively intended for the security plugin, and that maybe this would be better to build the functionality directly in the core. Would appreciate @reta's feedback on this. If we start pulling on this thread then maybe a ton of additional work would be required (like introducing some notion of "super user"), but this approach kind of feels like a work-around versus what we would actually like to build.

Reply:
> but on the surface this PR seems like we're exposing something that is pretty much exclusively intended for the security plugin, and that maybe this would be better to build the functionality directly in the core. 

Definitely worth some thinking, @cwperks what could be alternatives here? Could we not touch the plugins but for example enhance `IndexNameExpressionResolver` to return system indices?

Reply:
> why we use Map<String, Set<String>> and not SystemIndexDescriptor fe?

Because security plugin would only need the index patterns of the system indices reserved by plugins. It could also be [SystemIndexDescriptor](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/SystemIndexDescriptor.java#L43), but that's marked as internal.

> why we use Consumer instead of simpler hook like (like other APIs, onIndexModule, onNodeStarted, ...)?

Its returns a Consumer because there needs to be a mechanism to enforce that only one plugin can implement this interface. By returning a consumer instead of void, we can check if more than one plugin has implemented the interface.

Reply:
Thanks @cwperks , I think these both issues could be lifted:
 - we should have generic API (not for security plugin only)
 - I don't see any reasons to offload the implementation constraint to methods - if we allow 1 plugin, that would stop right after first `SystemIndexPlugin` discovery

Reply:
@reta with this change there can be many SystemIndexPlugins, but only one SystemIndexPlugin that implements `onSystemIndex`.

Same with ActionPlugin.getRestHandlerWrapper, only security can implement but there are many ActionPlugins installed. 

If we went with an approach like this to pass the system indices to the security plugin, I think there should be a mechanism in place so that only the security plugin is exposed to the registered system indices. 

@andrross I'll create a write up about the "superuser" in the security plugin to evaluate whether that should be ported into core. At the very least, that would mean a breaking change to opensearch settings because `plugins.security.authcz.admin_dn` would have to be renamed. Perhaps simple `admin_dn`? This value is a list of distinguished names which represent the principal extracted from an admin certificate which is presented with a web request.

Reply:
> @reta with this change there can be many SystemIndexPlugins, but only one SystemIndexPlugin that implements onSystemIndex.

OK, so let just shelve if for a moment and consider the alternatives, since it becomes really usable by `security` plugin and not beneficial to anyone else.

Reply:
@cwperks I think we should continue the discussion of how to build the feature you want into core (i.e. a system index is only accessible by the plugin that creates it).

But in the short term, could you do want you want with something as simple as a new method like `IndexNameExpressionResolver::systemIndexNames`? That would seem to fit reasonably well with the responsibilities of `IndexNameExpressionResolver` and would side step a lot of questions raised by the proposed new extension point here.